image_builder/image_builder/docker/builder.py

```diff
@@ -55,38 +56,48 @@ def build(
         images_num = len(job_type.template_paths)
         for image_index in range(images_num):
             progress = f'({image_index+1}/{images_num})'
-
-            with wrap_context(f'building base image {progress}'):
-                base_image = _build_base_image(config, job_type, image_index, deployment_id, metric_labels)
-
-            template_path = job_type.template_paths[image_index]
-            if template_path is None:
-                assert manifest.docker and manifest.docker.dockerfile_path, 'User-module Dockerfile manifest.docker.dockerfile_path is expected'
-                dockerfile_path = workspace / manifest.docker.dockerfile_path
-            else:
-                with wrap_context(f'templating Dockerfile {progress}'):
-                    dockerfile_path = workspace / f'.fatman-{image_index}.Dockerfile'
-                    racetrack_version = os.environ.get('DOCKER_TAG', 'latest')
-                    template_dockerfile(manifest, template_path, dockerfile_path, base_image,
-                                        git_version, racetrack_version, job_type.version, env_vars)
+            build_progress = f'({image_index*2+1}/{images_num*2})'  # 2 actual builds (base + final) per job's image
 
             try:
+                with wrap_context(f'building base image {progress}'):
+                    update_deployment_phase(config, deployment_id, f'building image {build_progress}')
+                    base_image = _build_base_image(
+                        config, job_type, image_index, deployment_id, metric_labels, build_progress,
+                    )
+
+                build_progress = f'({image_index*2+2}/{images_num*2})'
+                template_path = job_type.template_paths[image_index]
+                if template_path is None:
+                    assert manifest.docker and manifest.docker.dockerfile_path, 'User-module Dockerfile manifest.docker.dockerfile_path is expected'
```

Comment:
This build function is one hell of a function :] long, with loop, multiple if statements, exception handling... it has high cognitive complexity. Suggestion here is to split it to subfunctions.